### PR TITLE
[stdlib] Optional `start` and `length` arg for `String._strref_dangerous`

### PR DESCRIPTION
**Pull request by fknfilewalker**
Normally when one wants to create a truncated StringRef, a ptr to the String together with offsets need to be used. This is annoying and dangerous and could be done more safely during the `_strref_dangerous` call. This also adds bound guards

### CODE REVIEW

File: stdlib/src/builtin/string.mojo
Comment (laszlokindrat) [inline stdlib/src/builtin/string.mojo:1139]: Please make these keyword-only to force callers to be explicit. Also, please use `Optional[Int]` for `length`:
```suggestion
    fn _strref_dangerous(self, *, start: Int = 0, length: Optional[Int] = None) -> StringRef:
```

**Review by laszlokindrat — COMMENTED** [review]
Thanks for the patch! @JoeLoser Do you have thoughts on this?

**Comment by JoeLoser** [comment]
> Thanks for the patch! @JoeLoser Do you have thoughts on this?

In general, use of `_strref_dangerous` and friends will go away in favor of some of the `StringSlice` (safe string slice) work going on internally, right?  I need to look closer at it before I can give a good review here.

**Comment by laszlokindrat** [comment]
> > Thanks for the patch! @JoeLoser Do you have thoughts on this?
> 
> In general, use of `_strref_dangerous` and friends will go away in favor of some of the `StringSlice` (safe string slice) work going on internally, right? I need to look closer at it before I can give a good review here.

Yeah, although I think this could be useful in the meanwhile, and low risk since it's a hidden API anyway.

@fknfilewalker Could you please rebase after the next nightly, and change this patch to update the potential uses in `https://github.com/modularml/mojo/pull/2710` and maybe other places? That would give us a definitive picture of how much this actually helps.

File: stdlib/src/builtin/string.mojo
Comment (laszlokindrat) [inline stdlib/src/builtin/string.mojo:1233]: Does this not work?
```suggestion
        var _length = min(length or len(self), len(self) - start)
```

File: stdlib/src/builtin/string.mojo
Comment (laszlokindrat) [inline stdlib/src/builtin/string.mojo:1231]: If we have these safety checks in the semantics, it would be good to ensure that negative `start` and `length` values are also clamped/handled.

File: docs/changelog.md
Comment (laszlokindrat) [inline docs/changelog.md:370]: Since this is a private API, could you please remove it from the changelog?

**Review by laszlokindrat — APPROVED** [review]
Thank you! I have a few more requests, but LGTM otherwise!

File: stdlib/src/builtin/string.mojo
Comment (fknfilewalker) [reply stdlib/src/builtin/string.mojo:1231]: no, `min(length, len(_ref) - start)` also does not work. Seems like the automatic unpacking does not work.
I am getting
`candidate not viable: argument #0 cannot be converted from 'Optional[Int]' to 'Int'`

Comment (laszlokindrat) [reply stdlib/src/builtin/string.mojo]: To use `min`, you definitely need to unpack it manually, and I think that's the intended behavior. I don't know if we want this with `or`; the desired behavior is not obvious if `length` has a value of `0`. Anyway, thanks for checking!

**Comment by laszlokindrat** [comment]
Also, can you please update `String.{end,start}swith` (and any other places you're aware of) to use these?

**Comment by fknfilewalker** [comment]
> Also, can you please update `String.{end,start}swith` (and any other places you're aware of) to use these?

For now I only checked string.mojo and there are not many places where it is really helpful and since you want to replace it with `span` anyway maybe we should just drop this?

**Comment by laszlokindrat** [comment]
Is it really only one place where it's useful? I would have expected more. If you find one or two more places, I think we should land it, otherwise you're probably right, it's not pulling its weight.

**Comment by laszlokindrat** [comment]
@fknfilewalker Did you find any other uses for this? If not, I think I will just close it for now, and we can revive later if it becomes relevant again.

**Comment by ConnorGray** [comment]
@fknfilewalker Thanks for your work on this so far.

Regarding whether we want to move forward with this, I think at the moment this change seems a bit under-motivated. From looking at where this new functionality is used, I think every call site is an example of a place that we will eventually want to migrate to using the relatively new `StringSlice` type.

We really appreciate you taking the time to make this contribution, but I think we're going to hold off on accepting it.

There are other valuable and worthwhile directions for improvements to the string handling code that could be made and we'd be eager to accept—if you're interested in helping out we'd love to accept those kinds of changes 🙂.

For example, as mentioned, we could transition away from using `_strref_dangerous()` in the places you highlighted and instead start using `StringSlice`-based indexing.

**Comment by fknfilewalker** [comment]
I support this decision :)

**Pull request closed without merging**
